src/deco/deco.cpp

```diff
@@ -39,77 +30,40 @@ void ViewDecoration::calculate_resize_size(int &target_width,
 }
 
 void ViewDecoration::notify_view_activated(bool active) {
-    colors = active ? &options->colors.focused : &options->colors.unfocused;
-    damage();
+    surface_ref->set_active(active);
+    if (!is_hidden())
+        damage();
 }
 
 // NOTE: This also gets called when first decorating the view, so this is when
-// we first properly get notified of its geometry and initialize cached_region.
+// we first properly get notified of its geometry and initialize the surface
+// region.
 void ViewDecoration::notify_view_resized(wf::geometry_t view_geometry) {
     node->view->damage();
-    width = view_geometry.width;
-    height = view_geometry.height;
-    cached_region = calculate_region();
-    node->view->damage();
-}
-
-void ViewDecoration::hide(bool enable) {
-    hidden = enable;
-    // TODO: remove this code when
-    // https://github.com/WayfireWM/wayfire/pull/1187 gets released.
-    // It should be node->view->damage().
-    node->refresh_geometry();
-}
-
-bool ViewDecoration::is_hidden() const {
-    return hidden || node->view->fullscreen;
-}
-
-bool ViewDecoration::is_mapped() const { return mapped; }
-
-wf::point_t ViewDecoration::get_offset() {
-    if (is_hidden())
-        return {0, 0};
-    return {-options->border_width, -options->border_width};
+    surface_ref->set_size(wf::dimensions(view_geometry));
+    if (!is_hidden())
+        node->view->damage();
 }
 
-wf::dimensions_t ViewDecoration::get_size() const { return {width, height}; }
+bool ViewDecoration::is_hidden() const { return (bool)surface; }
 
-bool ViewDecoration::accepts_input(int32_t sx, int32_t sy) {
-    for (const auto &ss : subsurfs)
-        if (ss->contains_point({sx, sy}))
-            return true;
-    return false;
+void ViewDecoration::attach_surface() {
+    assert(surface);
+    node->view->add_subsurface(std::move(surface), false);
 }
 
-void ViewDecoration::simple_render(const wf::framebuffer_t &fb, int x, int y,
-                                   const wf::region_t &damage) {
-    if (is_hidden())
-        return;
-
-    const wf::region_t region = cached_region + wf::point_t{x, y};
-
-    OpenGL::render_begin(fb);
-    for (const auto &scissor : region &damage) {
-        fb.logic_scissor(wlr_box_from_pixman_box(scissor));
-
-        for (const auto &ss : subsurfs)
-            ss->render({x, y}, fb.get_orthographic_projection());
-    }
-
-    OpenGL::render_end();
+void ViewDecoration::detach_surface() {
+    assert(!surface);
+    surface = std::move(node->view->remove_subsurface(surface_ref));
```

Comment:
This line generates a warning when I compile with clang about the move being unnecessary and preventing copy elision.
```suggestion
    surface = node->view->remove_subsurface(surface_ref);
```